extensions/amp-fresh/0.1/amp-fresh-manager.js

```diff
@@ -14,20 +14,107 @@
  * limitations under the License.
  */
 
-import {fromClass} from '../../../src/service';
+import {addParamToUrl} from '../../../src/url';
+import {fromClassForDoc} from '../../../src/service';
+import {isExperimentOn} from '../../../src/experiments';
+import {whenDocumentReady} from '../../../src/document-ready';
+import {user} from '../../../src/log';
+import {xhrFor} from '../../../src/xhr';
 
 
+/** @const */
+const TAG = 'amp-fresh';
+
 /**
  * Manages registered amp-fresh components and does the xhr request
  * and update.
  */
-export class AmpFreshManager {}
+export class AmpFreshManager {
+
+  /**
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+
+    /** @type {!../../../src/service/ampdoc-impl.AmpDoc} */
+    this.ampdoc = ampdoc;
+
+    if (!isExperimentOn(this.win, TAG)) {
+      return;
+    }
+
+    /** @private @const {!Object<string, !./amp-fresh.AmpFresh>} */
+    this.ampFreshInstances_ = Object.create(null);
+
+    /**
+     * Used only for testing.
+     * @private @const {Promise<!Document>}
+     */
+    this.docPromise_ = this.fetchDocument_().then(doc => {
+      this.update_(doc);
+    }).catch(() => {
+      this.onFetchDocumentFailure_();
+    });
+  }
+
+  /**
+   * Registers an amp-fresh component instance to be managed by this windows
+   * amp-fresh-manager instance.
+   * @param {string} id
+   * @param {!./amp-fresh.AmpFresh} ampFreshIntance
+   */
+  register(id, ampFreshInstance) {
+    user().assert(!this.ampFreshInstances_[id],
+        `duplicate amp-fresh id used: ${id}`);
+    this.ampFreshInstances_[id] = ampFreshInstance;
+  }
+
+  /**
+   * @return {!Promise<!Document>}
+   * @private
+   */
+  fetchDocument_() {
+    const url = addParamToUrl(this.ampdoc.win.location.href,
+        'amp-fresh', '1');
+    return Promise.all([
+      xhrFor(this.ampdoc.win).fetchDocument(url),
+      this.ampdoc.whenReady(),
+    ]).then(args => args[0]);
+  }
+
+  /**
+   * @param {!Document} docFromServer
+   * @private
+   */
+  update_(docFromServer) {
+    Object.keys(this.ampFreshInstances_).forEach(id => {
+      const ampFresh = this.ampFreshInstances_[id];
+      const counterpart = docFromServer.querySelector(`#${id}`);
```

Comment:
mostly for test purposes. since you can only call getElementById on document while i pass in document fragments for tests. willing to change of course if you feel strongly since it is not a very convincing reason